_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+
+The context bound syntax remains unchanged.
+
+## Automatically converting values between types
+
+The implicit conversion feature is dangerous. For this reason, in Scala 3, the compiler
+will warn you every time it is used. You can disable the warning at your own risk by
+explicitly importing the feature into the current scope.
+
+The Java standard library provides an `Optional` type which is very similar to `Option`.
+If you are working with a Java library which produces a lot of objects with this type,
+but you also have many `Option`s around, you might want to define an automatic conversion.
+
+This is done by extending a new trait defined in the standard library:
+```scala
+abstract class Conversion[-T, +U] extends (T => U)
+```
+
+In Scala 3 you can define it in this way:
+
+```scala
+// OptionalConversion.scala
+import java.util.Optional
+
+object OptionalConversion:
+
+  given [A] as Conversion[Optional[A], Option[A]]:
+    def apply(in: Optional[A]): Option[A] =
+      if in.isPresent then Some(in.get())
+      else None
+```
+
+You can then use the syntax described above for the import:
+
+```scala
+import java.util.Optional
+import OptionalConversion.given
+
+@main def main =
+  val opt: Option[Int] = Optional.empty[Int]()
+```
+
+As mentioned above, the compiler will warn you about this dangerous feature:
+```
+Use of implicit conversion class given_Conversion_Optional_Option in object OptionalConversion should be enabled
+by adding the import clause 'import scala.language.implicitConversions'
+or by setting the compiler option -language:implicitConversions.
+See the Scala docs for value scala.language.implicitConversions for a discussion
+why the feature should be explicitly enabled.
+```
+The warning can be silenced by adding `import scala.language.implicitConversions`, as
+the message suggests.
+
+### Find out more
+
+You can learn more about this feature in the [documentation](https://dotty.epfl.ch/docs/reference/contextual/conversions.html).
+
+### How was this done in Scala 2?
+
+Scala 2 relied on `implicit` defs and `implicit` function values to implement this pattern:
+
+```scala
+import java.util.Optional
+
+object OptionalConversion {
+  implicit def optionalToOption[A](in: Optional[A]): Option[A] =
+    if (in.isPresent) Some(in.get())
+    else None
+}
+```
+
+which could then be imported:
+```scala
+import OptionalConversion._
+```
+
+Note that at a glance, it is not clear that the definition is intended as an automatic conversion to an expected type:
+ - the name of the method can be a hint, but this is merely a convention that
+ other developers might not share
+ - the implicit def may be intended give you extension methods on its parameter type, depending on its result
+ - there is not straightforward way to verify that this function will not be used as implicit
+ argument for a function with an implicit parameter list
```

Comment:
```suggestion
 - the name of the method can be a hint, but this is merely a convention that
   other developers might not share
 - the implicit def may be intended give you extension methods on its parameter type, depending on its result
 - there is not straightforward way to verify that this function will not be used as implicit
   argument for a function with an implicit parameter list
```
otherwise I'm not sure it will properly render